generator-star fails when function expressions have params

These test cases should be valid:

```js
{
    code: "var foo = function* ( arg1 ){};",
    args: [1, "start"],
    ecmaFeatures: { "generators": true }
},
{
    code: "var foo = function *( arg1 ){};",
    args: [1, "end"],
    ecmaFeatures: { "generators": true }
}
```

Here are the errors:

```
  1) generator-star var foo = function* ( arg1 ){};:
     AssertionError: Should have no errors but had 1: [ { ruleId: 'generator-star',
    severity: 1,
    node:
     { type: 'FunctionExpression',
       id: null,
       params: [Object],
       defaults: [],
       body: [Object],
       rest: null,
       generator: true,
       expression: false,
       range: [Object],
       loc: [Object],
       parent: [Object] },
    message: 'Expected no space before *.',
    line: 1,
    column: 10,
    source: 'function* ( arg1 ){}' } ]: expected 1 to equal 0

  2) generator-star var foo = function *( arg1 ){};:
     AssertionError: Should have no errors but had 1: [ { ruleId: 'generator-star',
    severity: 1,
    node:
     { type: 'FunctionExpression',
       id: null,
       params: [Object],
       defaults: [],
       body: [Object],
       rest: null,
       generator: true,
       expression: false,
       range: [Object],
       loc: [Object],
       parent: [Object] },
    message: 'Expected a space before *.',
    line: 1,
    column: 10,
    source: 'function *( arg1 ){}' } ]: expected 1 to equal 0
```

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/7874985-generator-star-fails-when-function-expressions-have-params?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/rules/generator-star.js
+++ b/lib/rules/generator-star.js
@@ -15,16 +15,23 @@ module.exports = function(context) {
     var position = context.options[0] || "end";
 
     /**
-     * Check the position of the start compared to the expected position.
-     * @param {Object} node - the entire function node
-     * @param {Object} starToken - the star token
+     * Check the position of the star compared to the expected position.
+     * @param {ASTNode} node - the entire function node
      * @returns {void}
      */
-    function checkStarPosition(node, starToken) {
+    function checkStarPosition(node) {
+        var starToken;
 
         if (!node.generator) {
             return;
         }
+
+        // Blocked, pending decision to fix or work around in eslint/espree#36
+        if (context.getAncestors().pop().method) {
+            return;
+        }
+
+        starToken = context.getFirstToken(node, 1);
 
         // check for function *name() {}
         if (position === "end") {
@@ -38,7 +45,7 @@ module.exports = function(context) {
         // check for function* name() {}
         if (position === "start") {
 
-            // * ends where the previous identifier ends
+            // * begins where the previous identifier ends
             if (starToken.range[0] !== context.getTokenBefore(starToken).range[1]) {
                 context.report(node, "Expected no space before *.");
             }
@@ -46,18 +53,8 @@ module.exports = function(context) {
     }
 
     return {
-
-        "FunctionDeclaration": function (node) {
-            var starToken = context.getTokenBefore(node.id);
-            checkStarPosition(node, starToken);
-        },
-
-        "FunctionExpression": function (node) {
-
-            // count back an extra token if you have a named anonymous function
-            var starToken = context.getTokenBefore(node.body, node.id ? 3 : 2);
-            checkStarPosition(node, starToken);
-        }
+        "FunctionDeclaration": checkStarPosition,
+        "FunctionExpression": checkStarPosition
     };
 
 };

--- a/tests/lib/rules/generator-star.js
+++ b/tests/lib/rules/generator-star.js
@@ -73,12 +73,27 @@ eslintTester.addRuleTest("lib/rules/generator-star", {
             ecmaFeatures: { "generators": true, "objectLiteralShorthandMethods": true }
         },
         {
+            code: "var x = {\n    *foo(){} }",
+            args: [1, "start"],
+            ecmaFeatures: { "generators": true, "objectLiteralShorthandMethods": true }
+        },
+        {
             code: "var foo = function*(){}",
             args: [1, "start"],
             ecmaFeatures: { "generators": true }
         },
         {
             code: "var foo = function*(){}",
+            args: [1, "end"],
+            ecmaFeatures: { "generators": true }
+        },
+        {
+            code: "var foo = function* ( arg1 ){};",
+            args: [1, "start"],
+            ecmaFeatures: { "generators": true }
+        },
+        {
+            code: "var foo = function *( arg1 ){};",
             args: [1, "end"],
             ecmaFeatures: { "generators": true }
         }

